pkg/chunked/storage_linux.go

```diff
@@ -1131,32 +1133,17 @@ func (c *chunkedDiffer) ApplyDiff(dest string, options *archive.TarOptions, diff
 	var convertedBlobSize int64
 
 	if c.convertToZstdChunked {
-		fd, err := unix.Open(dest, unix.O_TMPFILE|unix.O_RDWR|unix.O_CLOEXEC, 0o600)
+		payload, digester, err := c.requestWholeBlob()
 		if err != nil {
-			return graphdriver.DriverWithDifferOutput{}, &fs.PathError{Op: "open", Path: dest, Err: err}
+			return graphdriver.DriverWithDifferOutput{}, err
 		}
-		blobFile := os.NewFile(uintptr(fd), "blob-file")
 		defer func() {
-			if blobFile != nil {
-				blobFile.Close()
+			if payload != nil {
+				payload.Close()
 			}
 		}()
 
-		// calculate the checksum before accessing the file.
```

Comment:
Presumably we were doing this for a reason (security?) before...is it just that we think `convertTarToZstdChunked` can now safely operate on untrusted input?

Reply:
hmm I don't remember the security implications, @mtrmac do you think this approach is fine or should I just close this PR?

Reply:
On balance, I’d prefer this not to be merged as is.

---

This exposes the conversion code and decompression, to malicious input in more situations.

In principle, the conversion code and decompression must be able to handle malicious input _anyway_, because an attacker could, in many cases, refer to the malicious input in a valid manifest, without triggering this check.

But _some_ users might have a policy which only accepts signed images, i.e. the malicious input would only be digested, not processed otherwise.

I’m really more worried about the decompression than the chunked conversion. That’s a lot of bit manipulation, potentially in an unsafe language “for performance”, with a fair history of vulnerabilities.

… and for ordinary image pulls, we currently also stream the input through decompression, only validating the digest concurrently; we don’t validate the digest before starting to decompress. So, in that sense, we are already accepting a larger part of the risk.

So I don’t feel that strongly about it here.

---

More importantly, if this is aiming at the c/storage `ApplyDiff` = c/image `PutBlob` (not `PutBlobPartial`) path, c/image is currently creating a temporary file, _and verifying the digest_, on that path, without c/storage having any way to prevent it; and c/storage wouldn’t need to do this. (Also, c/image streams the data to the file, and digests it, concurrently for several layers, without holding any storage locks, which seems valuable.)

So, if the goal is _code structure_, not _performance or `PutBlobPartial`_, I don’t see that this makes any difference for the `ApplyDiff` path, and it is just a performance/risk trade-off for the `PutBlobPartial` path.

If you don’t actually _care_ about the performance at this point, we get the increased risk but not any benefit we value — so I’d prefer to close the PR and not merge this; we can always resurrect it later.

Except for the one line that allows `convertTarToZstdChunked` to accept an arbitrary `io.Reader`, and the missing `payload.Close`.